src/ngraph/attribute_visitor.hpp

```diff
@@ -27,137 +28,125 @@ namespace ngraph
         virtual void on_adapter(const std::string& name,
-                                ValueAccessor<std::vector<double>>& adapter)
-        {
-            on_adapter(name, static_cast<ValueAccessor<void>&>(adapter));
-        }
+                                ValueAccessor<std::vector<float>>& adapter);
         virtual void on_adapter(const std::string& name,
-                                ValueAccessor<std::vector<std::string>>& adapter)
-        {
-            on_adapter(name, static_cast<ValueAccessor<void>&>(adapter));
-        }
-        // Use an adapter for non-primitive types
-        template <typename T>
-        // typename std::enable_if<std::is_class<T>::value, void>::type
-        void on_attribute(const std::string& name, T& value)
-        {
-            AttributeAdapter<T> adapter(value);
-            on_adapter(name, adapter);
-        }
-        void on_attribute(const std::string& name, op::AutoBroadcastSpec& value)
-        {
-            AttributeAdapter<op::AutoBroadcastType> adapter(value.m_type);
-            on_adapter(name, adapter);
-        }
-        void on_attribute(const std::string& name, op::BroadcastModeSpec& value)
-        {
-            AttributeAdapter<op::BroadcastType> adapter(value.m_type);
-            on_adapter(name, adapter);
-        }
+                                ValueAccessor<std::vector<double>>& adapter);
+        virtual void on_adapter(const std::string& name,
+                                ValueAccessor<std::vector<std::string>>& adapter);
+        /// \brief Hook for adapters that need visitor access
+        virtual void on_adapter(const std::string& name, VisitorAdapter& adapter);
+
+        /// The generic visitor. There must be a definition of AttributeAdapter<T> that can convert
+        /// to a ValueAccessor<U> for one of the on_adpater methods.
+        template <typename AT>
+        void on_attribute(const std::string& name, AT& value)
+        {
+            AttributeAdapter<AT> adapter(value);
+            on_adapter(get_name_with_context(name), adapter);
+        }
+        /// \returns The nested context of visits
+        const std::vector<std::string>& get_context() const { return m_context; }
+        /// \returns context prepended to names
+        virtual std::string get_name_with_context(const std::string& name);
+        /// \brief Start visiting a nested structure
+        virtual void start_structure(const std::string& name);
+        /// \brief Finish visiting a nested structure
+        virtual void finish_structure();
+        using node_id_t = std::string;
+        static const node_id_t invalid_node_id;
+        /// \brief Associate a node with an id.
+        ///
+        /// No node may be used as an attribute unless it has already been registered with an ID.
+        /// References to nodes are visited with a ValueAccessor of their ID.
+        virtual void register_node(const std::shared_ptr<Node>& node, node_id_t id);
```

Comment:
Do we need to register all nodes for the TensorIterator?

Reply:
All nodes used in the body. If you deserialize them before TensorIterator and register them as you deserialize with whatever ID you use to refer to them in the body, everything will work.

Reply:
Is it means that I should detect operations in the TI body for each topology and register them? If yes it is not flexible.
How should I detect the set of operations that TI for some topology contains? AlsoTI can contain some custom operations.

Reply:
In the spec for TensorIterator, each layer in the body has an ID, so those are the IDs. How do you plan to deserialize those nodes in the body and connect their inputs? The same way you deserialize the other nodes in the graph, or does the visitor deserialize them? That is the question. It doesn't matter if the nodes are custom.

Reply:
Could you please provide a test or more detailed explanation which will illustrate the way how we can create the TI layer?

In order to be on the same page... At the current moment I really don't fully understand the way, how we can create TI from Visitors API.